SOLR-16889 Rate Limiter should stop processing on 429

stillalex opened the pull request:
https://issues.apache.org/jira/browse/SOLR-16889

<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Solr:

* https://issues.apache.org/jira/projects/SOLR

For something minor (i.e. that wouldn't be worth putting in release notes), you can skip JIRA. 
To create a Jira issue, you will need to create an account there first.

The title of the PR should reference the Jira issue number in the form:

* SOLR-####: <short description of problem or changes>

SOLR must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

PR for short cutting the request if Rate Limiter decides against processing. also takes care of cleaning up the request.

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

dsmiley commented on solr/core/src/java/org/apache/solr/servlet/ServletUtils.java:
```diff
@@ -216,16 +217,19 @@ static void rateLimitRequest(
     } finally {
+      if (needsCleanup) {
+        consumeInputFully(request, response);
+        SolrRequestInfo.reset();
+        SolrRequestParsers.cleanupMultipartFiles(request);
```
This looks wrong.... we won't reset SolrRequestInfo sometimes?  Why not?  Not doing so is dangerous; messing up access to SolrRequestInfo for future requests on the same thread.

And structurally this doesn't look right.  We shouldn't be adding additional callers to some of these methods.  Control flow should be well balanced logical and avoid calling methods from multiple spots under different conditions.  It's not that it's wrong necessarily but is hard to understand and reason about.

stillalex replied:
thank you for taking a look. I agree with the control flow idea 100%. it is very difficult to work around code that is so poorly set in place.
I introduced this flag because ` traceHttpRequestExecution2` does resource cleanup "sometimes", take a look (I actually think the 'else' block in there is dead code, but it's not relevant to this PR). 
ideally this entire block "rate limiter method calling into trace method calling into mystery method" would be wrapped into a try/finally with correct cleanup but it's not so I added this flag which was "best effort cleanup" in case the trace method does not get called.
Also, is there an actual test that verifies correct resource cleanup? I am happy to extend it with this change.

If you have stronger opinions (proposing a proper refactoring) and are good for a review, I am happy to dive deeper, but so far there are not too many reviewers available.

dsmiley replied:
I'm happy to review around this code (here I am) so restructuring I think is necessary first.  @gus-asf might -- he brought about I'd current structure.

stillalex replied:
thank you for offering. I will try to clean this up and at the same time reduce collateral changes as much as possible.

dsmiley commented on solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java:
```diff
@@ -209,8 +208,8 @@ public void doFilter(
     Tracer t = getCores() == null ? GlobalTracer.get() : getCores().getTracer();
     request.setAttribute(Tracer.class.getName(), t);
     RateLimitManager rateLimitManager = coreService.getService().getRateLimitManager();
-    request.setAttribute(RateLimitManager.class.getName(), rateLimitManager);
```
It's not clear to me if the RateLimitManager was on the request merely out of convenience for the existing logic or if it was intentionally used for miscellaneous unforeseen use-cases (and you are now removing it).  Shrug.

stillalex replied:
luckily we have very good tooling support to show where a class is used inside a project. I can 100% tell that this is not used anywhere else. Happy to be shown different. I think this was just a partial/unfinished plan to do something and it ended up in an awkward place where it is set as an attribute instead of simply passing the ref.
I took the liberty to remove some of the tech debt by cleaning up some code I saw was not clear, if this is putting too much burden on the review I will be happy to revert, let me know.

dsmiley replied:
Of course we can use Java tooling to look within Solr to see but this may be here for external / plugins.  In general we put objects on the request for plugins.  I'm not familiar with the RateLimitManager to consider if there would be a point for any plugin to want to get it from the request.  My intuition agrees with you.

stillalex replied:
hmm. good point. I was quick to say it's not used in Solr, but you are of course right plugins are something I forgot about. still, the rate limiter doesn't strike me as being a good fit for a request attribute. I am ok with leaving this in if you think there is a risk of regression.

stillalex commented:
Pushing a deeper refactoring. the cleanup will happen only once on a finally block inside the main entry method. I think the null tracer codepath is dead code, I don't see anywhere it could be null unless there is a bad test setup (so test must be fixed instead). in some sense this is needing a major split to servlet filters (rate limiting, tracing), but until that happens at least the code is clearer wrt. resource cleanup.

dsmiley commented on solr/core/src/java/org/apache/solr/servlet/ServletUtils.java:
```diff
@@ -197,35 +197,30 @@ static void configExcludes(PathExcluder excluder, String patternConfig) {
+      Thread.currentThread().interrupt();
+      throw new SolrException(ErrorCode.SERVER_ERROR, e.getMessage());
     } finally {
+      consumeInputFully(request, response);
```
Lets move these three lines somewhere else as this doesn't have to do with reate limiting (this method).  Perhaps generic doFilter.

dsmiley reviewed (approved):
Major improvement already!

stillalex commented:
as suggested, I moved the resource cleanup to `doFilter`. I also removed another layer of try/catch blocks.
please review carefully, I did my best to not introduce any bugs but I might have missed something in the process. thank you @dsmiley!

stillalex commented:
@dsmiley please let me know your thoughts on the latest version
